src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs

```diff
@@ -0,0 +1,454 @@
+        [InlineData(true, "")]
+        [InlineData(false, "")]
+        [InlineData(true, "Non-Empty")]
+        [InlineData(false, "Non-Empty")]
+        public async Task Read_FullLengthSynchronous_Success(bool transferEncodingChunked, string text)
+        {
+            byte[] expected = Encoding.UTF8.GetBytes(text);
+            Task<HttpListenerContext> contextTask = _listener.GetContextAsync();
+
+            using (HttpClient client = new HttpClient())
+            {
+                client.DefaultRequestHeaders.TransferEncodingChunked = transferEncodingChunked;
+                Task<HttpResponseMessage> clientTask = client.PostAsync(_factory.ListeningUrl, new StringContent(text));
+
+                HttpListenerContext context = await contextTask;
+                if (transferEncodingChunked)
+                {
+                    Assert.Equal(-1, context.Request.ContentLength64);
+                    Assert.Equal("chunked", context.Request.Headers["Transfer-Encoding"]);
+                }
+                else
+                {
+                    Assert.Equal(expected.Length, context.Request.ContentLength64);
+                    Assert.Null(context.Request.Headers["Transfer-Encoding"]);
+                }
+
+                byte[] buffer = new byte[expected.Length];
+                int bytesRead = context.Request.InputStream.Read(buffer, 0, buffer.Length);
+                Assert.Equal(expected.Length, bytesRead);
+                Assert.Equal(expected, buffer);
+
+                // Subsequent reads don't do anything.
+                Assert.Equal(0, context.Request.InputStream.Read(buffer, 0, buffer.Length));
+                Assert.Equal(expected, buffer);
+
+                context.Response.Close();
+                using (HttpResponseMessage response = await clientTask)
+                {
+                    Assert.Equal(200, (int)response.StatusCode);
+                }
+            }
+        }
+
+        [ConditionalTheory(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task Read_LargeLengthAsynchronous_Success(bool transferEncodingChunked)
+        {
+            string text = new string('a', 128 * 1024 + 1); // More than 128kb
+            byte[] expected = Encoding.UTF8.GetBytes(text);
+            Task<HttpListenerContext> contextTask = _listener.GetContextAsync();
+
+            using (HttpClient client = new HttpClient())
+            {
+                client.DefaultRequestHeaders.TransferEncodingChunked = transferEncodingChunked;
+                Task<HttpResponseMessage> clientTask = client.PostAsync(_factory.ListeningUrl, new StringContent(text));
+
+                HttpListenerContext context = await contextTask;
+
+                // If the size is greater than 128K, then we limit the size, and have to do multiple reads.
+                byte[] buffer = new byte[expected.Length];
```

Comment:
This limit is coming from http.sys, windows native stack that httplistener uses. We don't have this restriction in managed implementation. This will fail on Unix.